feat: Add expiration date to access token & hmac keys

CosDiabos opened the pull request:
**Description**
Implements custom date expiration for AccessTokens and Hmac Tokens leveraging the `expires` column within auth_identities table. 

Implements `setIdentityExpirationById()` model method, `updateAccessTokenExpiration()`, `removeAccessTokenExpiration()`, `isAccessTokenExpired()`, `hasAccessTokenExpiry()`, `updateHmacTokenExpiration()`, `removeHmacTokenExpiration()`, `isHmacTokenExpired()`, `hasHmacTokenExpiry()` trait methods and the CLI command `shield:hmac invalidateAll`.

Implements and closes #926 

**Checklist:**
- [x] Securely signed commits
- [x] Component(s) with PHPDoc blocks, only if necessary or adds value
- [x] Unit testing, with >80% coverage
- [x] User guide updated
- [x] Conforms to style guide

CosDiabos committed:
feat: add expiration date to access token & hmac keys/tokens.

CosDiabos committed:
fix test assertion

CosDiabos committed:
fix test for lower PHP version and PHPDoc blocks

michalsn commented on src/Authentication/Traits/HasAccessTokens.php:
```diff
@@ -34,15 +37,18 @@ trait HasAccessTokens
+     * @throws InvalidArgumentException
      */
-    public function generateAccessToken(string $name, array $scopes = ['*']): AccessToken
+    public function generateAccessToken(string $name, array $scopes = ['*'], ?string $expiresAt = null): AccessToken
```
The type chosen for `$expiresAt` is quite strange to me - why `?string` and not `?Time`?
`Time` seems a lot more universal and we wouldn't need `checkExpiresAtFormat()` method at all.

CosDiabos replied:
You are right, a ```string``` doesn't make sense in this situation, a ```Time``` type is the correct answer, specially, as you mentioned gives people more flexibility. What I do think happen was, I got caught up with the date/time format and then, mistakenly went with strings without re-evaluating.

michalsn commented on src/Authentication/Traits/HasAccessTokens.php:
```diff
@@ -165,4 +171,64 @@ public function setAccessToken(?AccessToken $accessToken): self
+     *
+     * @throws InvalidArgumentException
+     */
+    public function getAccessTokenTimeToExpire(?AccessToken $accessToken, string $format = 'date'): string|null
```
Relying directly on the `Time` object seems more universal to me. Let the developers decide what format they want to display the expiration date in. I would get rid of this method.

CosDiabos replied:
Like I've mentioned in my previous comment, I agree with you and will change it accordingly. The idea with this method initially, was to return the "humanized" representation easily but then for the sake of completeness, I also brought in the date which now makes more sense getting both from the Time object directly.

michalsn commented on src/Authentication/Traits/HasAccessTokens.php:
```diff
@@ -165,4 +171,64 @@ public function setAccessToken(?AccessToken $accessToken): self
+     *
+     * @return bool|null Returns true if Access Token has expired, false if not, and null if the expire field is null
+     */
+    public function hasAccessTokenExpired(?AccessToken $accessToken): bool|null
```
Can we return here only `true`/`false` - this seems more appropriate.

We can introduce another method, like `canAccessTokenExpire()` to determine whether the token has an expiration date.

CosDiabos replied:
Yes, that makes more sense.

michalsn reviewed (commented):
Sorry you had to wait so long for feedback.

The PR looks nice, although I have questions about the reasons for some of the "implementation" choices.

CosDiabos committed:
fix token expiration type, tests and docs references.
fix invalidateAll hmac cli command.
implements CanAccessTokenExpire() and CanHmacTokenExpire() methods.
removes hasAccessTokenExpired() and hasHmacTokenExpired() methods.

CosDiabos committed:
fix file

CosDiabos committed:
fix namespace case.
fix hmac invalidateAll() expires field.

datamweb commented:
Is it possible to generate a token without a time limit?

CosDiabos commented:
> Is it possible to generate a token without a time limit?

Sure! Just skip that argument when generating a token. It will follow the defined token lifetime though.

michalsn commented on docs/references/authentication/hmac.md:
```diff
@@ -97,6 +97,59 @@ You can revoke all HMAC Keys with the `revokeAllHmacTokens()` method.
+
+The following support methods are also available:
+
+`hasHmacTokenExpired(AccessToken $HmacToken)` - Checks if the HMAC key has expired. Returns `true` if the HMAC key has expired, `false` if not, and `null` if the expire date is null.
```
```suggestion
`hasHmacTokenExpired(AccessToken $HmacToken)` - Checks if the HMAC key has expired. Returns `true` if the HMAC key has expired, `false` if not.
```

michalsn commented on docs/references/authentication/hmac.md:
````diff
@@ -97,6 +97,59 @@ You can revoke all HMAC Keys with the `revokeAllHmacTokens()` method.
+```php
+// Expiration date = 2024-11-03 12:00:00
+$expiresAt = Time::parse('2024-11-03 12:00:00');
+$token = $this->user->generateHmacToken('foo', ['foo:bar'], $expiresAt);
````
```suggestion
$token = $this->user->generateHmacToken('foo', ['foo.bar'], $expiresAt);
```
Scope names should avoid using a colon, according to the docs.

michalsn commented on docs/references/authentication/hmac.md:
```diff
@@ -97,6 +97,59 @@ You can revoke all HMAC Keys with the `revokeAllHmacTokens()` method.
+$token = $user->setHmacTokenExpirationById($token->id, $expiresAt);
+
+// Or Expiration date = 1 month + 15 days into the future
+$expiresAt = Time::now();
+$expiresAt = $expiresAt->addMonths(1);
+$expiresAt = $expiresAt->addDays(15);
```
```suggestion
$expiresAt = Time::now()->addMonths(1)->addDays(15);
```

michalsn commented on docs/references/authentication/hmac.md:
````diff
@@ -97,6 +97,59 @@ You can revoke all HMAC Keys with the `revokeAllHmacTokens()` method.
+
+```php
+$expiresAt = Time::parse('2024-11-03 12:00:00');
+$token = $this->user->generateHmacToken('foo', ['foo:bar'], $expiresAt);
````
```suggestion
$token = $this->user->generateHmacToken('foo', ['foo.bar'], $expiresAt);
```

michalsn commented on docs/references/authentication/hmac.md:
````diff
@@ -97,6 +97,59 @@ You can revoke all HMAC Keys with the `revokeAllHmacTokens()` method.
+```php
+$expiresAt = Time::parse('2024-11-03 12:00:00');
+
+$token = $this->user->generateHmacToken('foo', ['foo:bar'], $expiresAt);
````
```suggestion
$token = $this->user->generateHmacToken('foo', ['foo.bar'], $expiresAt);
```

michalsn commented on docs/references/authentication/hmac.md:
````diff
@@ -228,6 +281,9 @@ that CodeIgniter provides.
 public $unusedTokenLifetime = YEAR;
 ```
 
+### HMAC Keys Expiration vs Lifetime
+Expiration and Lifetime are different concepts. The lifetime is the maximum time allowed for the HMAC Key to exist since its last use. HMAC Key expiration, on the other hand, is a set date in which the HMAC Key will cease to function.
````
```suggestion
### HMAC Keys Expiration vs Lifetime

Expiration and Lifetime are different concepts. The lifetime is the maximum time allowed for the HMAC Key to exist since its last use. HMAC Key expiration, on the other hand, is a set date in which the HMAC Key will cease to function.
```
Just formatting - one line space between the heading and the content.

michalsn commented on docs/references/authentication/tokens.md:
````diff
@@ -137,6 +137,60 @@ that CodeIgniter provides.
+```php
+// Expiration date = 2024-11-03 12:00:00
+$expiresAt = Time::parse('2024-11-03 12:00:00');
+$token = $this->user->generateAccessToken('foo', ['foo:bar'], $expiresAt);
````
```suggestion
$token = $this->user->generateAccessToken('foo', ['foo.bar'], $expiresAt);
```

michalsn commented on docs/references/authentication/tokens.md:
```diff
@@ -137,6 +137,60 @@ that CodeIgniter provides.
+$user->setAccessTokenExpirationById($token->id, $expiresAt);
+
+// Or Expiration date = 1 month + 15 days into the future
+$expiresAt = Time::now();
+$expiresAt = $expiresAt->addMonths(1);
+$expiresAt = $expiresAt->addDays(15);
```
```suggestion
$expiresAt = Time::now()->addMonths(1)->addDays(15);
```

michalsn commented on docs/references/authentication/tokens.md:
```diff
@@ -137,6 +137,60 @@ that CodeIgniter provides.
+
+The following support methods are also available:
+
+`hasAccessTokenExpired(AccessToken $accessToken)` - Checks if Access Token has expired. Returns `true` if the Access Token has expired, `false` if not, and `null` if the expire date is not set.
```
```suggestion
`hasAccessTokenExpired(AccessToken $accessToken)` - Checks if Access Token has expired. Returns `true` if the Access Token has expired, `false` if not.
```

michalsn commented on docs/references/authentication/tokens.md:
````diff
@@ -137,6 +137,60 @@ that CodeIgniter provides.
+```php
+$expiresAt = Time::parse('2024-11-03 12:00:00');
+
+$token = $this->user->generateAccessToken('foo', ['foo:bar'], $expiresAt);
````
```suggestion
$token = $this->user->generateAccessToken('foo', ['foo.bar'], $expiresAt);
```

michalsn commented on docs/references/authentication/tokens.md:
````diff
@@ -137,6 +137,60 @@ that CodeIgniter provides.
+```php
+$expiresAt = Time::parse('2024-11-03 12:00:00');
+
+$token = $this->user->generateAccessToken('foo', ['foo:bar'], $expiresAt);
````
```suggestion
$token = $this->user->generateAccessToken('foo', ['foo.bar'], $expiresAt);
```

michalsn commented on src/Authentication/Traits/HasAccessTokens.php:
```diff
@@ -165,4 +171,46 @@ public function setAccessToken(?AccessToken $accessToken): self
+     *
+     * @return bool Returns true if Access Token has expired, false if not
+     */
+    public function hasAccessTokenExpired(?AccessToken $accessToken): bool
```
```suggestion
    public function hasAccessTokenExpired(AccessToken $accessToken): bool
```
I don't think we should allow null.

CosDiabos replied:
Yeah, that make sense.

I also noticed that  `setAccessToken()` and `setHmacToken()` are allowing a null value, and they probably shouldn't unless there is a reason to clear the access/hmac token.

michalsn replied:
Well, the `$accessToken` in `setAccessToken()` and `setHmacToken()` can be `null`, because the user may simply have no token (e.g. see `AccessTokens::loginById()`). 

But this new method doesn't rely on the `$this->currentAccessToken` property, but is rather intended to be used to administer tokens when we use `accessTokens()`. So, my thinking was the `$accessToken` should not be optional.

michalsn commented on src/Authentication/Traits/HasAccessTokens.php:
```diff
@@ -165,4 +171,46 @@ public function setAccessToken(?AccessToken $accessToken): self
+     */
+    public function canAccessTokenExpire(AccessToken $accessToken): bool
+    {
+        return isset($accessToken->expires);
```
```suggestion
        return $accessToken->expires !== null;
```
Since the existence of the object is guaranteed, we can go with a strict check.

michalsn commented on src/Authentication/Traits/HasHmacTokens.php:
```diff
@@ -156,4 +161,46 @@ public function setHmacToken(?AccessToken $accessToken): self
+     *
+     * @return bool Returns true if Access Token has expired, false if not
+     */
+    public function hasHmacTokenExpired(?AccessToken $accessToken): bool
```
```suggestion
    public function hasHmacTokenExpired(AccessToken $accessToken): bool
```

michalsn commented on src/Authentication/Traits/HasAccessTokens.php:
```diff
@@ -165,4 +171,46 @@ public function setAccessToken(?AccessToken $accessToken): self
+     *
+     * @return bool Returns true if expiration date is set or updated.
+     */
+    public function setAccessTokenExpirationById(int $id, Time $expiresAt): bool
```
I wonder... since we already allow the expiration date to be changed, should we also allow the expiration to be removed altogether?

CosDiabos replied:
I think it could be useful. Right now, if a expiration date is set, it can't be undone intuitively.

Would it be better to make its own method or clear it when it isn't present (`?Time` as you mentioned)?

michalsn replied:
`?Time` is enough - I don't think we need a separate method for this. But this will require changes also to the `setIdentityExpirationById()` method.

CosDiabos replied:
Great, sounds good!

michalsn commented on src/Authentication/Traits/HasHmacTokens.php:
```diff
@@ -156,4 +161,46 @@ public function setHmacToken(?AccessToken $accessToken): self
+     *
+     * @return bool Returns true if expiration date is set or updated.
+     */
+    public function setHmacTokenExpirationById(int $id, Time $expiresAt): bool
```
Same question as with `setAccessTokenExpirationById()` - should we allow `?Time`?

michalsn commented on src/Authentication/Traits/HasHmacTokens.php:
```diff
@@ -156,4 +161,46 @@ public function setHmacToken(?AccessToken $accessToken): self
+     */
+    public function canHmacTokenExpire(AccessToken $accessToken): bool
+    {
+        return isset($accessToken->expires);
```
```suggestion
        return $accessToken->expires !== null;
```

michalsn reviewed (changes requested):
@CosDiabos Thank you for the updates. I have a few other comments.

Of other things, I see that one action "Detect Merge Commits" reports an error. I'm not sure because of which commit, but you can try to rebase (there were some changes in the codebase since you started this PR). Or alternatively, if you identify the commit that is interfering, you can try to remove it from this PR - here are quite clear instructions that should help: https://graphite.dev/guides/how-to-delete-a-git-commit#deleting-old-or-specific-commits

@datamweb - is this failed action a deal breaker for us or not really?

datamweb commented:
> is this failed action a deal breaker for us or not really?

@michalsn , commit [e6c10c0d65c3360639fe39120d48e3949b740238](https://github.com/codeigniter4/shield/pull/1219/commits/e6c10c0d65c3360639fe39120d48e3949b740238) was correctly detected by the action. The user has likely used `pull` to update their branch, whereas we emphasize using `rebase` instead.  

We prefer **avoiding merge commits** to keep the *graph clean*, but it is *not mandatory*. However, based on your explanation, I believe the user should be able to handle the issue.  

If the user needs to update their branch, they should use `rebase` instead.
https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/workflow.md#updating-your-branch

CosDiabos committed:
fix hasAccessTokenExpired()& hasHmacTokenExpired() args.
fix docs refinement.
fix test.

CosDiabos commented:
Hey!

Thank you all for your help and support.

Yes I had merged because I had a couple of out of date files that I was working on. I've removed the merge commit with the rebase. I think now everything is in order with most change requests updated!

paulbalandan commented on docs/references/authentication/tokens.md:
```diff
@@ -125,7 +125,7 @@ Configure **app/Config/AuthToken.php** for your needs.
 ### Access Token Lifetime
 
-Tokens will expire after a specified amount of time has passed since they have been used.
+Tokens will expire after a specified amount of time has passed since they last have been used.
```
```suggestion
Tokens will expire after a specified amount of time has passed since they have last been used.
```

michalsn commented on docs/references/authentication/tokens.md:
```diff
@@ -137,6 +137,58 @@ that CodeIgniter provides.
+
+
+### Access Token Expiration vs Lifetime
+Expiration and Lifetime are different concepts. The lifetime is the maximum time allowed for the token to exist since its last use. Token expiration, on the other hand, is a set date in which the Access Token will cease to function.
```
Please add a line break between the section header and the content.

michalsn reviewed (commented):
As for Psalm errors, this is out of the scope of this PR. Seems like the new release has changed some default settings. You can add `ensureOverrideAttribute="false"` to the `psalm.xml` file, right after `findUnusedCode="false"` to fix this.

CosDiabos commented:
> As for Psalm errors, this is out of the scope of this PR. Seems like the new release has changed some default settings. You can add `ensureOverrideAttribute="false"` to the `psalm.xml` file, right after `findUnusedCode="false"` to fix this.

Yeah, I skimmed the errors looking for a file that I could have edited but couldn't find it. Thanks for the tip.

CosDiabos committed:
fix token expiration args, support remove expiration.
fix tokens / hmac docs.
fix psalm xml attribute: ensureOverrideAttribute.

michalsn reviewed (approved):
Thank you - it looks good. Let's see what others think.

datamweb commented on docs/references/authentication/hmac.md:
```diff
@@ -97,6 +97,60 @@ You can revoke all HMAC Keys with the `revokeAllHmacTokens()` method.
+
+By default, the HMAC keys don't expire unless they meet the HMAC Keys lifetime expiration after their last used date.
+
+HMAC keys can be set to expire through the `generateHmacToken()` method. This takes the expiration date as the $expiresAt argument. To update or remove an existing HMAC key expiration date use `setHmacTokenExpirationById($HmacTokenID, $expiresAt)`
```
```suggestion
HMAC keys can be set to expire through the `generateHmacToken()` method. This takes the expiration date as the `$expiresAt` argument. To update or remove an existing HMAC key expiration date use `setHmacTokenExpirationById($HmacTokenID, $expiresAt)`
```

datamweb commented on docs/references/authentication/hmac.md:
```diff
@@ -97,6 +97,60 @@ You can revoke all HMAC Keys with the `revokeAllHmacTokens()` method.
+
+HMAC keys can be set to expire through the `generateHmacToken()` method. This takes the expiration date as the $expiresAt argument. To update or remove an existing HMAC key expiration date use `setHmacTokenExpirationById($HmacTokenID, $expiresAt)`
+
+`$expiresAt` [Time](/libraries/time.html) object
```
```suggestion
`$expiresAt` [Time](https://codeigniter.com/user_guide/libraries/time.html) object
```

datamweb commented on docs/references/authentication/hmac.md:
````diff
@@ -97,6 +97,60 @@ You can revoke all HMAC Keys with the `revokeAllHmacTokens()` method.
+```php
+// Expiration date = 2024-11-03 12:00:00
+$expiresAt = Time::parse('2024-11-03 12:00:00');
+$token = $this->user->generateHmacToken('foo', ['foo:bar'], $expiresAt);
````
```suggestion
$token = $this->user->generateHmacToken('foo', ['foo.bar'], $expiresAt);
```

datamweb commented on docs/references/authentication/hmac.md:
```diff
@@ -97,6 +97,60 @@ You can revoke all HMAC Keys with the `revokeAllHmacTokens()` method.
+
+By default, the HMAC keys don't expire unless they meet the HMAC Keys lifetime expiration after their last used date.
+
+HMAC keys can be set to expire through the `generateHmacToken()` method. This takes the expiration date as the $expiresAt argument. To update or remove an existing HMAC key expiration date use `setHmacTokenExpirationById($HmacTokenID, $expiresAt)`
```
_To update or remove an existing HMAC key expiration date use `setHmacTokenExpirationById($HmacTokenID, $expiresAt)`_

Since these are two distinct responsibilities, I think it's better to create separate functions for updating and removing the HMAC token expiration. This improves code clarity and maintainability.  

Update expiration date:
   ```php
   public function updateHmacTokenExpiration($hmacTokenID,  $expiresAt): bool
   ```
Remove expiration date (set to null)
   ```php
   public function removeHmacTokenExpiration($hmacTokenID): bool
   ```

What do you think?

kenjis replied:
Yes, I also think `update***()` and `remove***()` would be better.

CosDiabos replied:
Yeah, michalsn and I talked about that same question (separation) earlier and kept everything on the same method.

However, I understand that the difference in functionality should warrant their own methods, also makes the removal of the expiration more conscious and less error prone action. I agree in splitting the actions.

datamweb commented on docs/references/authentication/hmac.md:
```diff
@@ -97,6 +97,60 @@ You can revoke all HMAC Keys with the `revokeAllHmacTokens()` method.
+
+The following support methods are also available:
+
+`hasHmacTokenExpired(AccessToken $HmacToken)` - Checks if the HMAC key has expired. Returns `true` if the HMAC key has expired, `false` if not.
```
Wouldn't it be better to use `isHmacTokenExpired` instead of `hasHmacTokenExpired`? 
This name is clearer and more naturally indicates that the function checks whether the token has expired.

CosDiabos replied:
I agree it's better semantically.

datamweb commented on docs/references/authentication/hmac.md:
````diff
@@ -97,6 +97,60 @@ You can revoke all HMAC Keys with the `revokeAllHmacTokens()` method.
+$this->user->hasHmacTokenExpired($token); // Returns true
+```
+
+`canHmacTokenExpire(AccessToken $HmacToken)` - Checks if HMAC key has an expiration set. Returns `true` or `false` accordingly.
````
Wouldn't `hasHmacTokenExpiry` be a better choice instead of `canHmacTokenExpire`? 
The word **hasExpiry** more clearly indicates whether the token has an expiration set, avoiding potential confusion with can, which might imply permission or the ability to change.

datamweb commented on docs/references/authentication/hmac.md:
````diff
@@ -97,6 +97,60 @@ You can revoke all HMAC Keys with the `revokeAllHmacTokens()` method.
+```
+
+You can also easily set all existing HMAC keys/tokens as expired with the `spark` command:
+```
````
```suggestion
```console
```

datamweb commented on docs/references/authentication/hmac.md:
````diff
@@ -97,6 +97,60 @@ You can revoke all HMAC Keys with the `revokeAllHmacTokens()` method.
+```
+php spark shield:hmac invalidateAll
+```
+**Careful!** This command invalidates _all_ keys for _all_ users.
````
```suggestion
**Careful!** This command invalidates _all_ keys for _all_ users.

!!! warning

    This command invalidates _all_ keys for _all_ users.
```

datamweb commented on docs/references/authentication/tokens.md:
```diff
@@ -137,6 +137,61 @@ that CodeIgniter provides.
+
+By default, the Access Tokens don't expire unless they meet the Access Token lifetime expiration after their last used date.
+
+Access Tokens can be set to expire through the `generateAccessToken()` method. This takes the expiration date as the $expiresAt argument. To update or remove an existing HMAC key expiration date use `setAccessTokenById($HmacTokenID, $expiresAt)`
```
```suggestion
Access Tokens can be set to expire through the `generateAccessToken()` method. This takes the expiration date as the `$expiresAt` argument. To update or remove an existing HMAC key expiration date use `setAccessTokenById($HmacTokenID, $expiresAt)`
```
I have the same question about the function name `setHmacTokenExpirationById` here as well. What do you think? 😃

datamweb commented on docs/references/authentication/tokens.md:
```diff
@@ -137,6 +137,61 @@ that CodeIgniter provides.
+
+Access Tokens can be set to expire through the `generateAccessToken()` method. This takes the expiration date as the $expiresAt argument. To update or remove an existing HMAC key expiration date use `setAccessTokenById($HmacTokenID, $expiresAt)`
+
+`$expiresAt` [Time](/libraries/time.html) object
```
```suggestion
`$expiresAt` [Time](https://codeigniter.com/user_guide/libraries/time.html) object
```

datamweb commented on docs/references/authentication/tokens.md:
````diff
@@ -137,6 +137,61 @@ that CodeIgniter provides.
+$this->user->hasAccessTokenExpired($token); // Returns true
+```
+
+`canAccessTokenExpire(AccessToken $HmacToken)` - Checks if Access Token has an expiration set. Returns `true` or `false` accordingly.
````
What do you think? 😃
```suggestion
`hasAccessTokenExpiry(AccessToken $HmacToken)` - Checks if Access Token has an expiration set. Returns `true` or `false` accordingly.
```

datamweb commented on docs/references/authentication/tokens.md:
```diff
@@ -137,6 +137,61 @@ that CodeIgniter provides.
+
+The following support methods are also available:
+
+`hasAccessTokenExpired(AccessToken $accessToken)` - Checks if Access Token has expired. Returns `true` if the Access Token has expired, `false` if not.
```
What do you think? 😃
```suggestion
`isAccessTokenExpired(AccessToken $accessToken)` - Checks if Access Token has expired. Returns `true` if the Access Token has expired, `false` if not.
```

datamweb commented on src/Authentication/Authenticators/AccessTokens.php:
```diff
@@ -154,6 +154,19 @@ public function check(array $credentials): Result
+        if (
+            $token->expires !== null
+            && $token->expires->isBefore(
+                Time::now(),
```
Did the tools add this, or is it extra?
```suggestion
                Time::now()
```

CosDiabos replied:
Yes it's from the CS tool. Due to 'trailing_comma_in_multiline' rule, I guess.

datamweb commented on src/Models/UserIdentityModel.php:
```diff
@@ -224,6 +229,24 @@ public function getAllAccessTokens(User $user): array
+        $this->checkUserId($user);
+
+        return $this->where('user_id', $user->id)
+            ->where('type', $type_token)
```
If the `$type_token` is `null`, what happens with the `where('type', $type_token) `condition?

CosDiabos replied:
Actually, it shouldn't be accepting a null value for that argument because `$type_token` should always be defined (Access or Hmac).

But, what I'm realising is if the `$id` is already provided, the `$type_token` arg and ` ->where('type', $type_token)` on the update seems too much since the `$id` already points there.

datamweb commented on tests/Authentication/HasAccessTokensTest.php:
```diff
@@ -140,24 +141,111 @@ public function testTokenCanNoTokenSet(): void
+        $this->assertSame($tokenExpiration->format('Y-m-d h:i:s'), $this->user->currentAccessToken()->expires->format('Y-m-d h:i:s'));
+
+        $tokenExpiration = $tokenExpiration->addMonths(1);
+        $tokenExpiration = $tokenExpiration->addYears(1);
```
```suggestion
        $tokenExpiration = $tokenExpiration->addYears(1)->addMonths(1);
```

datamweb commented on tests/Authentication/HasHmacTokensTest.php:
```diff
@@ -149,10 +150,110 @@ public function testHmacTokenCantNoTokenSet(): void
+
+        $this->assertSame($tokenExpiration->format('Y-m-d h:i:s'), $this->user->currentHmacToken()->expires->format('Y-m-d h:i:s'));
+
+        $tokenExpiration = $tokenExpiration->addMonths(1);
+        $tokenExpiration = $tokenExpiration->addYears(1);
```
```suggestion
        $tokenExpiration = $tokenExpiration->addYears(1)->addMonths(1);
```

datamweb commented on tests/Authentication/HasHmacTokensTest.php:
```diff
@@ -149,10 +150,110 @@ public function testHmacTokenCantNoTokenSet(): void
+     */
+    public function testHmacTokenTimeToExpired(): void
+    {
+        $tokenExpiration = Time::now();
+        $tokenExpiration = $tokenExpiration->addYears(1);
```
```suggestion
        $tokenExpiration = Time::now()->addYears(1);
```

datamweb commented on tests/Authentication/HasHmacTokensTest.php:
```diff
@@ -149,10 +150,110 @@ public function testHmacTokenCantNoTokenSet(): void
+     */
+    public function testHmacTokenRemoveExpiration(): void
+    {
+        $tokenExpiration = Time::now();
+        $tokenExpiration = $tokenExpiration->addYears(1);
```
```suggestion
        $tokenExpiration = Time::now()->addYears(1);
```

datamweb reviewed (changes requested):
@CosDiabos Thank you for submitting the PR, and I apologize for the long delay. I have provided some suggested changes.  

Regarding the method names, my understanding is based on the documentation, but I might have misinterpreted it. If I have misunderstood any part of the code, I would appreciate your clarification.

kenjis commented on src/Authentication/Traits/HasHmacTokens.php:
```diff
@@ -35,17 +38,19 @@ trait HasHmacTokens
-     * @param list<string> $scopes Permissions the token grants
+     * @param string       $name      Token name
+     * @param list<string> $scopes    Permissions the token grants
+     * @param Time         $expiresAt Expiration date
```
```suggestion
     * @param Time|null    $expiresAt Expiration date
```

kenjis commented on src/Authentication/Traits/HasAccessTokens.php:
```diff
@@ -34,15 +37,18 @@ trait HasAccessTokens
-     * @param list<string> $scopes Permissions the token grants
+     * @param string       $name      Token name
+     * @param list<string> $scopes    Permissions the token grants
+     * @param Time         $expiresAt Expiration date
```
```suggestion
     * @param Time|null    $expiresAt Expiration date
```

CosDiabos committed:
fix tokens expiration method names.
fix docs and tests.

datamweb commented on docs/references/authentication/hmac.md:
```diff
@@ -146,10 +146,12 @@ $this->user->canHmacTokenExpire($token2); // Returns false
-**Careful!** This command invalidates _all_ keys for _all_ users.
+
+!!! warning
+    This command invalidates _all_ keys for _all_ users.
```
```suggestion

    This command invalidates _all_ keys for _all_ users.
```

CosDiabos committed:
fix hmac docs warning section

datamweb commented on docs/references/authentication/hmac.md:
```diff
@@ -97,6 +97,62 @@ You can revoke all HMAC Keys with the `revokeAllHmacTokens()` method.
+$expiresAt = Time::parse('2024-11-03 12:00:00');
+$token = $this->user->generateHmacToken('foo', ['foo.bar'], $expiresAt);
+
+$this->user->hasHmacTokenExpired($token); // Returns true
```
Did you mean?
```suggestion
$this->user->isHmacTokenExpired($token); // Returns true
```

datamweb commented on docs/references/authentication/hmac.md:
```diff
@@ -97,6 +97,63 @@ You can revoke all HMAC Keys with the `revokeAllHmacTokens()` method.
+$expiresAt = Time::parse('2024-11-03 12:00:00');
+
+$token = $this->user->generateHmacToken('foo', ['foo.bar'], $expiresAt);
+$this->user->canHmacTokenExpire($token); // Returns true
```
And ?

datamweb commented on docs/references/authentication/hmac.md:
```diff
@@ -97,6 +97,63 @@ You can revoke all HMAC Keys with the `revokeAllHmacTokens()` method.
+$this->user->canHmacTokenExpire($token); // Returns true
+
+$token2 = $this->user->generateHmacToken('bar');
+$this->user->canHmacTokenExpire($token2); // Returns false
```
And?

datamweb commented on docs/references/authentication/tokens.md:
```diff
@@ -137,6 +137,61 @@ that CodeIgniter provides.
+
+$token = $this->user->generateAccessToken('foo', ['foo.bar'], $expiresAt);
+
+$this->user->hasAccessTokenExpired($token); // Returns true
```
And?
```suggestion
$this->user->isAccessTokenExpired($token); // Returns true
```

datamweb commented on docs/references/authentication/tokens.md:
```diff
@@ -137,6 +137,61 @@ that CodeIgniter provides.
+$expiresAt = Time::parse('2024-11-03 12:00:00');
+
+$token = $this->user->generateAccessToken('foo', ['foo.bar'], $expiresAt);
+$this->user->canAccessTokenExpire($token2); // Returns false
```
And?
```suggestion
$this->user->canAccessTokenExpire($token2); // Returns false
```

datamweb commented on docs/references/authentication/tokens.md:
```diff
@@ -137,6 +137,61 @@ that CodeIgniter provides.
+$this->user->canAccessTokenExpire($token2); // Returns false
+
+$token2 = $this->user->generateAccessToken('bar');
+$this->user->canAccessTokenExpire($token); // Returns true
```
And?
```suggestion
$this->user->canAccessTokenExpire($token); // Returns true
```

CosDiabos committed:
fix hmac expiration docs sample

paulbalandan commented on src/Authentication/Traits/HasHmacTokens.php:
```diff
@@ -156,4 +160,67 @@ public function setHmacToken(?AccessToken $accessToken): self
+    /**
+     * Checks if the provided Access Token is expired.
+     *
+     * @return bool Returns true if Access Token is expired, false if not
```
This comment is unnecessary as it can be inferred from the method description what it does.
```suggestion
```

paulbalandan commented on src/Authentication/Traits/HasAccessTokens.php:
```diff
@@ -165,4 +171,67 @@ public function setAccessToken(?AccessToken $accessToken): self
+    /**
+     * Checks if the provided Access Token is expired.
+     *
+     * @return bool Returns true if Access Token is expired, false if not
```
```suggestion
```

datamweb reviewed (approved):
Please update the PR description. Also, be aware that, based on coordination with the team, others may provide suggestions for improvements here. I appreciate your patience in advance.

paulbalandan commented on src/Authentication/Traits/HasAccessTokens.php:
```diff
@@ -165,4 +171,67 @@ public function setAccessToken(?AccessToken $accessToken): self
+    }
+
+    /**
+     * Checks if the current Hmac token can expire
```
```suggestion
     * Checks if the access token has a set expiration date
```

paulbalandan commented on src/Authentication/Traits/HasAccessTokens.php:
```diff
@@ -165,4 +171,67 @@ public function setAccessToken(?AccessToken $accessToken): self
+    /**
+     * Checks if the current Hmac token can expire
+     *
+     * @return bool Returns true if AccessToken can expire.
```
```suggestion
```

paulbalandan commented on src/Authentication/Traits/HasHmacTokens.php:
```diff
@@ -156,4 +160,67 @@ public function setHmacToken(?AccessToken $accessToken): self
+    }
+
+    /**
+     * Checks if the current Hmac token can expire
```
```suggestion
     * Checks if the current HMAC token has a set expiration date
```

paulbalandan commented on src/Authentication/Traits/HasHmacTokens.php:
```diff
@@ -156,4 +160,67 @@ public function setHmacToken(?AccessToken $accessToken): self
+    /**
+     * Checks if the current Hmac token can expire
+     *
+     * @return bool Returns true if Hmac Token can expire.
```
```suggestion
```

paulbalandan commented on src/Commands/Hmac.php:
```diff
@@ -196,4 +202,33 @@ static function ($identity) use ($uIdModelSub, $encrypter, $that): void {
+        $uIdModel    = new UserIdentityModel();
+        $uIdModelSub = new UserIdentityModel();
+
+        $that = $this;
```
Why do you need to assign `$this` to a variable? You can just make the closure below as not static then just access `$this->write` normally.

CosDiabos replied:
Hey!

Thanks! I wasn't sure about how the Closure worked, and the docs sample for `chunk()` uses a static method, which, seemed the right way to do it.

paulbalandan commented on src/Commands/Hmac.php:
```diff
@@ -196,4 +202,33 @@ static function ($identity) use ($uIdModelSub, $encrypter, $that): void {
+
+        $uIdModel->where('type', 'hmac_sha256')->orderBy('id')->chunk(
+            100,
+            static function ($identity) use ($uIdModelSub, $that): void {
```
```suggestion
            function ($identity) use ($uIdModelSub): void {
```

paulbalandan commented on src/Commands/Hmac.php:
```diff
@@ -196,4 +202,33 @@ static function ($identity) use ($uIdModelSub, $encrypter, $that): void {
+        $uIdModel->where('type', 'hmac_sha256')->orderBy('id')->chunk(
+            100,
+            static function ($identity) use ($uIdModelSub, $that): void {
+                $timeNow = Time::now(); // Current date/time
```
```suggestion
                $timeNow = Time::now();
```

paulbalandan commented on src/Commands/Hmac.php:
```diff
@@ -196,4 +202,33 @@ static function ($identity) use ($uIdModelSub, $encrypter, $that): void {
+                $timeNow = Time::now(); // Current date/time
+
+                if (null !== $identity->expires && $identity->expires->isBefore($timeNow)) {
+                    $that->write('Hmac Key/Token ID: ' . $identity->id . ', already expired, skipped.');
```
```suggestion
                    $this->write('Hmac Key/Token ID: ' . $identity->id . ', already expired, skipped.');
```

paulbalandan commented on src/Commands/Hmac.php:
```diff
@@ -196,4 +202,33 @@ static function ($identity) use ($uIdModelSub, $encrypter, $that): void {
+                $identity->expires = $timeNow;
+                $uIdModelSub->save($identity);
+
+                $that->write('Hmac Key/Token ID: ' . $identity->id . ', set as expired.');
```
```suggestion
                $this->write('Hmac Key/Token ID: ' . $identity->id . ', set as expired.');
```

CosDiabos commented:
> Please update the PR description. Also, be aware that, based on coordination with the team, others may provide suggestions for improvements here. I appreciate your patience in advance.

Updated! No problem!

CosDiabos committed:
fix comments and docs

CosDiabos committed:
fix test assertion

paulbalandan commented on docs/references/authentication/hmac.md:
```diff
@@ -97,6 +97,63 @@ You can revoke all HMAC Keys with the `revokeAllHmacTokens()` method.
 
+## Expiring HMAC Keys
+
+By default, the HMAC keys don't expire unless they meet the HMAC Keys lifetime expiration after their last used date.
```
```suggestion
By default, the HMAC keys don't expire unless they reach the HMAC keys' lifetime expiration after their last use date.
```

paulbalandan commented on docs/references/authentication/hmac.md:
```diff
@@ -97,6 +97,63 @@ You can revoke all HMAC Keys with the `revokeAllHmacTokens()` method.
+
+By default, the HMAC keys don't expire unless they meet the HMAC Keys lifetime expiration after their last used date.
+
+HMAC keys can be set to expire through the `generateHmacToken()` method. This takes the expiration date as the `$expiresAt` argument. To update an existing HMAC key expiration date use `updateHmacTokenExpiration($hmacTokenID, $expiresAt)` and to remove `removeHmacTokenExpiration($hmacTokenID)`.
```
```suggestion
HMAC keys can be set to expire through the `generateHmacToken()` method. This takes the expiration date as the `$expiresAt` argument. To update an existing HMAC key expiration date, use `updateHmacTokenExpiration($hmacTokenID, $expiresAt)`. To remove it, use `removeHmacTokenExpiration($hmacTokenID)`.
```

paulbalandan commented on docs/references/authentication/hmac.md:
```diff
@@ -97,6 +97,63 @@ You can revoke all HMAC Keys with the `revokeAllHmacTokens()` method.
+
+The following support methods are also available:
+
+`isHmacTokenExpired(AccessToken $HmacToken)` - Checks if the HMAC key is expired. Returns `true` if the HMAC key is expired, `false` if not.
```
```suggestion
`isHmacTokenExpired(AccessToken $hmacToken)` - Checks if the HMAC key is expired. Returns `true` if the HMAC key is expired; otherwise, returns `false`.
```

paulbalandan commented on docs/references/authentication/hmac.md:
````diff
@@ -97,6 +97,63 @@ You can revoke all HMAC Keys with the `revokeAllHmacTokens()` method.
+$this->user->isHmacTokenExpired($token); // Returns true
+```
+
+`hasHmacTokenExpiry(AccessToken $HmacToken)` - Checks if HMAC key has an expiration set. Returns `true` or `false` accordingly.
````
```suggestion
`hasHmacTokenExpiry(AccessToken $hmacToken)` - Checks if HMAC key has an expiration set. Returns `true` if the HMAC key is expired; otherwise, returns `false`.
```

paulbalandan commented on docs/references/authentication/hmac.md:
```diff
@@ -228,6 +285,10 @@ that CodeIgniter provides.
 
+### HMAC Keys Expiration vs Lifetime
+
+Expiration and Lifetime are different concepts. The lifetime is the maximum time allowed for the HMAC Key to exist since its last use. HMAC Key expiration, on the other hand, is a set date in which the HMAC Key will cease to function.
```
```suggestion
Expiration and lifetime are two different concepts. The lifetime is the maximum time allowed for the HMAC Key to exist since its last use. HMAC Key expiration, on the other hand, is a set date in which the HMAC Key will cease to function.
```

paulbalandan commented on docs/references/authentication/tokens.md:
```diff
@@ -137,6 +137,61 @@ that CodeIgniter provides.
+
+## Expiring Access Tokens
+
+By default, the Access Tokens don't expire unless they meet the Access Token lifetime expiration after their last used date.
```
```suggestion
By default, the Access Tokens don't expire unless they reach the Access Token's lifetime expiration after their last use date.
```

paulbalandan commented on docs/references/authentication/tokens.md:
```diff
@@ -137,6 +137,61 @@ that CodeIgniter provides.
+
+By default, the Access Tokens don't expire unless they meet the Access Token lifetime expiration after their last used date.
+
+Access Tokens can be set to expire through the `generateAccessToken()` method. This takes the expiration date as the `$expiresAt` argument. To update an existing HMAC key expiration date use `updateAcessTokenExpiration($hmacTokenID, $expiresAt)` and to remove `removeAccessTokenExpiration($hmacTokenID)`.
```
```suggestion
Access Tokens can be set to expire through the `generateAccessToken()` method. This takes the expiration date as the `$expiresAt` argument. To update an existing HMAC key expiration date, use `updateAcessTokenExpiration($accessTokenID, $expiresAt)`. To remove it, use `removeAccessTokenExpiration($accessTokenID)`.
```

paulbalandan commented on docs/references/authentication/tokens.md:
```diff
@@ -137,6 +137,61 @@ that CodeIgniter provides.
+
+The following support methods are also available:
+
+`isAccessTokenExpired(AccessToken $accessToken)` - Checks if Access Token is expired. Returns `true` if the Access Token is expired, `false` if not.
```
```suggestion
`isAccessTokenExpired(AccessToken $accessToken)` - Checks if Access Token is expired. Returns `true` if the Access Token is expired; otherwise, returns `false`.
```

paulbalandan commented on docs/references/authentication/tokens.md:
````diff
@@ -137,6 +137,61 @@ that CodeIgniter provides.
+$this->user->isAccessTokenExpired($token); // Returns true
+```
+
+`hasAccessTokenExpiry(AccessToken $HmacToken)` - Checks if Access Token has an expiration set. Returns `true` or `false` accordingly.
````
```suggestion
`hasAccessTokenExpiry(AccessToken $accessToken)` - Returns `true` if the Access Token has a set expiration date; otherwise, returns `false`.
```

paulbalandan commented on docs/references/authentication/tokens.md:
```diff
@@ -137,6 +137,61 @@ that CodeIgniter provides.
+
+### Access Token Expiration vs Lifetime
+
+Expiration and Lifetime are different concepts. The lifetime is the maximum time allowed for the token to exist since its last use. Token expiration, on the other hand, is a set date in which the Access Token will cease to function.
```
```suggestion
Expiration and lifetime are two different concepts. The lifetime is the maximum time allowed for the token to exist since its last use. Token expiration, on the other hand, is a set date in which the Access Token will cease to function.
```

paulbalandan commented on src/Authentication/Traits/HasHmacTokens.php:
```diff
@@ -156,4 +160,63 @@ public function setHmacToken(?AccessToken $accessToken): self
+    public function isHmacTokenExpired(AccessToken $accessToken): bool
+    {
+        return $accessToken->expires !== null && $accessToken->expires->isBefore(Time::now());
+    }
```
```suggestion
     * Checks if the provided HMAC Token is expired.
     */
    public function isHmacTokenExpired(AccessToken $hmacToken): bool
    {
        return $hmacToken->expires !== null && $hmacToken->expires->isBefore(Time::now());
    }
```

paulbalandan commented on src/Authentication/Traits/HasHmacTokens.php:
```diff
@@ -156,4 +160,63 @@ public function setHmacToken(?AccessToken $accessToken): self
+    /**
+     * Sets an expiration for HMAC token by ID.
+     *
+     * @param int  $id        AccessToken ID
```
```suggestion
     * @param int  $id        HMAC Token ID
```

paulbalandan commented on src/Authentication/Traits/HasHmacTokens.php:
```diff
@@ -156,4 +160,63 @@ public function setHmacToken(?AccessToken $accessToken): self
+    /**
+     * Removes the expiration date for HMAC token by ID.
+     *
+     * @param int $id AccessToken ID
```
```suggestion
     * @param int $id HMAC Token ID
```

paulbalandan commented on src/Authentication/Traits/HasHmacTokens.php:
```diff
@@ -156,4 +160,63 @@ public function setHmacToken(?AccessToken $accessToken): self
+    public function hasHmacTokenExpiry(AccessToken $accessToken): bool
+    {
+        return $accessToken->expires !== null;
+    }
```
```suggestion
    public function hasHmacTokenExpiry(AccessToken $hmacToken): bool
    {
        return $hmacToken->expires !== null;
    }
```

paulbalandan reviewed (commented):
Some grammar fixes

CosDiabos committed:
fix docs and comments.
fix updateHmacTokenExpiration() token refresh.

michalsn commented on src/Authentication/Traits/HasAccessTokens.php:
```diff
@@ -165,4 +171,63 @@ public function setAccessToken(?AccessToken $accessToken): self
+    /**
+     * Checks if the access token has a set expiration date
+     */
+    public function hasAccessTokenExpiry(AccessToken $accessToken): bool
```
I'm not sure if this has been changed or if it was here from the beginning, but reading the code, I thought: "Wait a minute, don't we have two methods with the same functionality?".

If someone reads the method name without checking its implementation, they might assume it returns true when the token is already expired, not just when an expiry date is set.

IMO "Has expiry" could be interpreted as "Has the token expired?" instead of "Does it have an expiry date?".

Maybe we could go with `canAccessTokenExpire()` or `isAccessTokenExpirable()`?

Same thing for Hmac.

CosDiabos replied:
The method names have been a challenge so far, haha.

> 
> Maybe we could go with `canAccessTokenExpire()` or `isAccessTokenExpirable()`?
> 

I don't think `hasAccessTokenExpiry()` is necessarily confusing, but `canAccessTokenExpire()` seems to be more straightforward. `isAccessTokenExpirable()` is confusing, imo

michalsn replied:
> The method names have been a challenge so far, haha.

Yeah, apparently 😆 

Let's wait a day - if there are no more opinions on the subject, choose what you think is best.

samsonasik commented on src/Authentication/Authenticators/AccessTokens.php:
```diff
@@ -154,6 +154,19 @@ public function check(array $credentials): Result
 
+        // Is expired ?
+        if (
+            $token->expires !== null
```
can this use instanceof, eg : 

```suggestion
            $token->expires instanceof Time
```

michalsn replied:
Sounds good.

michalsn replied:
@CosDiabos Would you find time for these final changes? I'm looking forward to merging this PR. Thanks!

CosDiabos committed:
fix tokens expire methods names.
fix check() expires type check.
updated docs & tests.

michalsn commented on src/Authentication/Traits/HasAccessTokens.php:
```diff
@@ -165,4 +171,63 @@ public function setAccessToken(?AccessToken $accessToken): self
+    /**
+     * Checks if the access token has a set expiration date
+     */
+    public function CanAccessTokenExpire(AccessToken $accessToken): bool
```
```suggestion
    public function canAccessTokenExpire(AccessToken $accessToken): bool
```
On more issue. The same applies to HMAC + all the mentions in the docs.

CosDiabos committed:
fix name case

michalsn commented:
@CosDiabos Thank you for your updates and patience, and congratulations on your first contribution to the project!

CosDiabos commented:
@michalsn You're welcome! Thank you all for your help and reviews throughout.
